ruby/ext/google/protobuf_c/protobuf.c

```diff
@@ -276,17 +276,22 @@ static void ObjectCache_Init(VALUE protobuf) {
   rb_const_set(protobuf, rb_intern("SIZEOF_VALUE"), INT2NUM(SIZEOF_VALUE));
 }
 
-VALUE ObjectCache_TryAdd(const void *key, VALUE val) {
+static VALUE ObjectCache_GetKey(const void *key) {
   VALUE key_val = (VALUE)key;
   PBRUBY_ASSERT((key_val & 3) == 0);
-  return rb_funcall(weak_obj_cache, item_try_add, 2, LL2NUM(key_val), val);
+  // Avoid overflow on 32-bit systems by discarding the bottom zeros
```

Comment:
I think the issue isn't overflow, but in https://github.com/ruby/ruby/blob/abd15ac775d41e6485f728fe0fad4cddf138d3ec/include/ruby/internal/arithmetic/long_long.h#L75-L86 it appears that if the value can fit as a `Fixnum` it will be stored that way. I suspect that on a 32-bit system, the value doesn't fit, so it has to be allocated on a heap via `Bignum`.

Example values:

0xff5b4af8 => 4284173048
(0xff5b4af8 >> 2) => 1071043262

Reply:
Can you add an unit test or an assert so that we can't regress this in the future? Most code paths aren't picky about a Fixnum vs a Bignum, but this one does.

Reply:
I've confirmed that this is indeed a Bignum vs. Fixnum issue. I've added an assertion, though the assertion will only be triggered on a 32-bit system with `NDEBUG` compile flag omitted. The entire test suite fails at the moment even without this assertion.